### PR TITLE
A few simple analogs from base

### PR DESCRIPTION
**Pull request by Divesh-Otwani**
I'm getting a head start on some of the analogs from base that we need.

I made a few minor design choices that I hope are straightforward :)

### CODE REVIEW

File: src/Data/Bool/Linear.hs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  , (&&)
+  , (||)
+  , not
+  )
+  where
+
+import Prelude (Bool(..))
+
+-- | True iff both are True.
+(&&) :: Bool #-> Bool #-> Bool
```
Comment (aspiwack) [inline src/Data/Bool/Linear.hs:18]: I'm wondering out loud whether we should take the opportunity to make a boolean algebra type class to abstract over these? Maybe not, it many not be useful often enough to be featured in a standard library.

File: src/Data/Ord/Linear.hs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+class Eq a => Ord a where
+  {-# MINIMAL compare | (<=) #-}
+  compare :: a #-> a #-> Ordering
+  compare = Unsafe.toLinear2 compare'
+    where
+      compare' :: Ord a => a -> a -> Ordering
+      compare' x y
+        | x == y = EQ
+        | x <= y = LT
+        | True = GT
```
Comment (aspiwack) [inline src/Data/Ord/Linear.hs:28]: If you need `Unsafe.toLinear`, then it's probably unsafe :wink: . 

The conclusion is that there is no linear definition of `compare` based on just `(==)` and `(<=)`, and that's fine.
```suggestion
```

File: src/Data/Ord/Linear.hs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  where
+
+import Data.Eq.Linear
+import qualified Prelude as Unrestricted
+import Prelude (Bool(..))
+import Data.Ord (Ordering(..))
+import qualified Unsafe.Linear as Unsafe
+
+class Eq a => Ord a where
+  {-# MINIMAL compare | (<=) #-}
```
Comment (aspiwack) [inline src/Data/Ord/Linear.hs:20]: As a consequence:

```suggestion
  {-# MINIMAL (<=) #-}
```

File: src/Data/Ord/Linear.hs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        GT -> x
+        _ -> y
+
+  min :: a #-> a #-> a
+  min = Unsafe.toLinear2 min'
+    where
+      min' :: Ord a => a -> a -> a
+      min' x y = case compare x y of
+        LT -> x
+        _ -> y
```
Comment (aspiwack) [inline src/Data/Ord/Linear.hs:59]: You can define these without `Unsafe.toLinear`, at the cost of one extra binding in `where`. Please do.

File: src/Data/Ord/Linear.hs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+instance Unrestricted.Ord a => Ord a where
+  compare = Unsafe.toLinear2 Unrestricted.compare
```
Comment (aspiwack) [inline src/Data/Ord/Linear.hs:62]: This is very wrong. We could give some (unsafe!) deriving-via combinator to that effect. But let's do that later: add an issue instead.

```suggestion
```

File: src/Data/Bool/Linear.hs (hunk 1 above)
Comment (facundominguez) [reply src/Data/Bool/Linear.hs:18]: Do this functions deserve a word about being strict? (unlike their non-linear variants)

Comment (aspiwack) [reply src/Data/Bool/Linear.hs:18]: This is a good point. Well-spotted.

Comment (Divesh-Otwani) [reply src/Data/Bool/Linear.hs:18]: Yes linearity forces them to be strict. That's interesting -- so in some sense giving linear types to functions in `base` that are linear does impact performance. (So a non-linear function using the linear `&&` non-linearly might have a different performance. So linear types don't impact the performance of a function that's already written, but they do force constraints on the implementation which might impact performance. This is an interesting observation.)

File: src/Data/Ord/Linear.hs (hunk 4 above)
Comment (Divesh-Otwani) [reply src/Data/Ord/Linear.hs:59]: I'm probably missing something but I don't think that's possible.

If you don't know `a` is `Dupable` as soon as you use it in `<` or `<=` you can't return it.

**Pull request merged**